contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.9.0) (metatx/ERC2771Forwarder.sol)
+
+pragma solidity ^0.8.19;
+
+import "../utils/cryptography/ECDSA.sol";
+import "../utils/cryptography/EIP712.sol";
+import "../utils/Nonces.sol";
+
+/**
+ * @dev An implementation of a production-ready forwarder compatible with ERC2771 contracts.
+ *
+ * This forwarder operates on forward requests that include:
+ *
+ * * `from`: An address to operate on behalf of. It is required to be equal to the request signer.
+ * * `to`: The address that should be called.
+ * * `value`: The amount of ETH to attach with the requested call.
+ * * `gas`: The amount of gas limit that will be forwarded with the requested call.
+ * * `nonce`: A unique transaction ordering identifier to avoid replayability and request invalidation.
+ * * `deadline`: A timestamp after which the request is not executable anymore.
+ * * `data`: Encoded `msg.data` to send with the requested call.
+ */
+contract ERC2771Forwarder is EIP712, Nonces {
+    using ECDSA for bytes32;
+
+    struct ForwardRequest {
```

Comment:
There's some redundancy in this struct. It doesn't need to be the same as the contents of the typehash. Since this impacts the size and cost of calldata that the relayer will submit when invoking `forward`, it makes sense to try to reduce the size of the struct.

The one thing that I see that can be removed is the nonce, because it's implicit with `from`. Given `from`, we can fetch the expected nonce from `Nonces`.